pkg/frontend/querymiddleware/astmapper/pruning.go

```diff
@@ -0,0 +1,181 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/cortexproject/cortex/blob/master/pkg/querier/astmapper/shard_summer.go
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: The Cortex Authors.
+
+package astmapper
+
+import (
+	"context"
+	"math"
+
+	"github.com/go-kit/log"
+	"github.com/prometheus/prometheus/promql/parser"
+)
+
+func NewQueryPruner(ctx context.Context, logger log.Logger) ASTMapper {
+	pruner := newQueryPruner(ctx, logger)
+	return NewASTExprMapper(pruner)
+}
+
+type queryPruner struct {
+	ctx    context.Context
+	logger log.Logger
+}
+
+func newQueryPruner(ctx context.Context, logger log.Logger) *queryPruner {
+	return &queryPruner{
+		ctx:    ctx,
+		logger: logger,
+	}
+}
+
+func (pruner *queryPruner) MapExpr(expr parser.Expr) (mapped parser.Expr, finished bool, err error) {
+	if err := pruner.ctx.Err(); err != nil {
+		return nil, false, err
+	}
+
+	switch e := expr.(type) {
+	case *parser.BinaryExpr:
+		return pruner.pruneBinOp(e)
+	default:
+		return e, false, nil
+	}
+}
+
+func (pruner *queryPruner) pruneBinOp(expr *parser.BinaryExpr) (mapped parser.Expr, finished bool, err error) {
+	switch expr.Op {
+	case parser.MUL:
+		return pruner.handleMultiplyOp(expr), false, nil
+	case parser.GTR, parser.LSS:
+		return pruner.handleCompOp(expr), false, nil
+	case parser.LOR:
```

Comment:
Might be something for another PR: we can do something similar for `and` and `unless`:

* if either side of an `and` is empty, then the whole op will return no results
* if the left side of an `unless` is empty, then the whole op will return no results
* if the right side of an `unless` is empty, then we can just use the left side and ignore the right side